### DISCUSSION TITLE
request_validation_exception_handler fails with NaN values

### DISCUSSION THREAD

**Discussion by Arnasst** [fastapi/fastapi#10141] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [x] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
from pydantic import BaseModel


class StringModel(BaseModel):
    value: str

app = FastAPI()

@app.post("/string")
async def post_str(
    body: StringModel
):
    return "Works well"
```

### Description

* Send a POST request to /string with body `{ "value" : NaN }`
* Receive `ValueError: Out of range float values are not JSON compliant`

The error seems to be happening when creating `JSONResponse` after `RequestValidationError` happens. By default json.dumps does not allow nan.
```
def render(self, content: typing.Any) -> bytes:
        return json.dumps(
            content,
            ensure_ascii=False,
            allow_nan=False,
            indent=None,
            separators=(",", ":"),
        ).encode("utf-8")
```

More detailed error path: 
```
  File "/opt/homebrew/lib/python3.11/site-packages/starlette/middleware/errors.py", line 162, in __call__
    await self.app(scope, receive, _send)
  File "/opt/homebrew/lib/python3.11/site-packages/starlette/middleware/exceptions.py", line 88, in __call__
    response = await handler(request, exc)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/fastapi/exception_handlers.py", line 23, in request_validation_exception_handler
    return JSONResponse(
           ^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/starlette/responses.py", line 196, in __init__
    super().__init__(content, status_code, headers, media_type, background)
  File "/opt/homebrew/lib/python3.11/site-packages/starlette/responses.py", line 55, in __init__
    self.body = self.render(content)
                ^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/starlette/responses.py", line 199, in render
    return json.dumps(
           ^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.4_1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/json/__init__.py", line 238, in dumps
    **kw).encode(obj)
          ^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.4_1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/json/encoder.py", line 200, in encode
    chunks = self.iterencode(o, _one_shot=True)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.4_1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/json/encoder.py", line 258, in iterencode
    return _iterencode(o, 0)
```

### Operating System

macOS

### FastAPI Version

0.101.1

### Pydantic Version

2.3.0

### Python Version

3.11

**Comment by JavierSanchezCastro** [comment]
It seems weird, I see this issue https://github.com/pydantic/pydantic/issues/1779 and I adapted it to your code

(Consider the example as a workaround)
```python3
from fastapi import FastAPI
from pydantic import BaseModel, field_validator
from math import isnan

class StringModel(BaseModel):
    value: str
    
    @field_validator("value", mode="before")
    def change_nan_to_none(cls, value):
        if type(value) is float and isnan(value):
            return None
        return value

app = FastAPI()

@app.post("/string")
async def post_str(body: StringModel):
    return "Works well"
```

Also, you can write this, and None values are going to be accepted (NaN included)
```value: str | None```

I read some other issues talking about JSONResponse and ORJSONResponse but nothing I tried works. I say again it seems weird and I'm not really sure if it's fastapi or pydantic side.

**Reply by JavierSanchezCastro** [reply]
This is also related #6330 

I write a comment in that discussion https://github.com/tiangolo/fastapi/discussions/8029

**Reply by Arnasst** [reply]
Yeah, I saw those issues and figured that they seem similar and "fixable" with changing the `default_response_class`, but that option doesn't affect error handler response classes.

**Comment by Arnasst** [comment]
Sorry, I forgot to add that my expected behaviour would be that such request would return 422, similarly to how when None or not a string is provided.

Thanks for the suggestion @JavierSanchezCastro, that's a good workaround for now, but adding this `@field_validator` to every string property in every Model seems slightly inconvenient 😄 

I'd think it's an error from FastAPI side, because pydantic successfully raises an error while validating model and the problem happens when FastAPI error handler catches it and tries to format the original request body (which contains NaN) into `JSONResponse`.

**Comment by Arnasst** [comment]
Another possible workaround in case there are a larger number of str properties and models is to override FastAPI request_validation_exception_handler to use ORJSONResponse: 
```python
from fastapi import FastAPI, Request, status
from fastapi.encoders import jsonable_encoder
from fastapi.exceptions import RequestValidationError
from fastapi.responses import ORJSONResponse
from pydantic import BaseModel


class StringModel(BaseModel):
    value: str

app = FastAPI()


@app.post("/string")
async def post_str(
    body: StringModel
):
    return "Works well"

@app.exception_handler(RequestValidationError)
def validation_exception_handler(_: Request, exc: RequestValidationError) -> ORJSONResponse:
    return ORJSONResponse(
        status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
        content=jsonable_encoder({"detail": exc.errors()}),
    )
```

This could also be a hint towards how this could be changed in FastAPI.
In there it would be possible to either allow other response classes or just modify the JSONResponse class to allow_nan, so these errors wouldn't appear.
